File: short-encoding-of-words.cpp
/*
 * @Description: https://leetcode-cn.com/problems/short-encoding-of-words/
 * @Author: 鱼摆摆
 * @Email: fishercloud@qq.com
 * @Github: https://github.com/FisherCloud
 * @Date: 2019-09-08 14:28:03
 * @LastEditTime: 2019-09-08 16:46:58
 */
#include <iostream>
#include <string>
#include <vector>
#include <set>

using namespace std;

class Solution
{
public:
    int minimumLengthEncoding(vector<string> &words)
    { // 参考官方题解
#if 1
        // 存储后缀
        set<string> good(words.begin(), words.end());
        for (auto &&word : words)
        {
            for (int i = 1; i < word.size(); i++)
            {
                good.erase(word.substr(i));
            }
        }

        int ans = 0;
        for (auto &&word : good)
        {
            ans += word.size() + 1;
        }

        return ans;
#else
// 字典树
#endif
    }
};

int main(int argc, char const *argv[])
{
    vector<string> words = {"time", "me", "bell"};
    Solution s;
    cout << s.minimumLengthEncoding(words);
    return 0;
}

File: problems/rs/short-encoding-of-words.rs
// https://leetcode-cn.com/problems/short-encoding-of-words/
#[derive(Debug)]
struct Solution {
}

impl Solution {
    pub fn minimum_length_encoding(words: Vec<String>) -> i32 {
        let mut res = 0i32;
        let mut words: Vec<String> =words.into_iter().map(|word| word.chars().rev().collect::<String>()).collect();
        words.sort();
        (0..words.len()).for_each(|i| {
            if i+1<words.len() && words[i+1].starts_with(&words[i]){
                res+=0;
            }else{
                res+=words[i].len() as i32+1;
            }
        });
        res
    }
}
